File: codes/ruby/chapter05/answer_A30.rb
def power(base, exponent, mod)
  power_value = base
  result = 1
  for i in 0...30
    divisor = 2 ** i
    if (exponent / divisor) % 2 == 1
      result = (result * power_value) % mod 
    end
    power_value = (power_value * power_value) % mod
  end
  return result
end

def division(numerator, denominator, mod)
  return (numerator * power(denominator, mod - 2, mod)) % mod
end

# Input
n, r = gets.split.map(&:to_i)
m = 1000000007

# Step 1: Calculate numerator
numerator = 1
for i in 1..n
  numerator = (numerator * i) % m
end

# Step 2: Calculate denominator
denominator = 1
for i in 1..r
  denominator = (denominator * i) % m
end
for i in 1..n-r
  denominator = (denominator * i) % m
end

# Step 3: Calculate the result
puts division(numerator, denominator, m)

File: codes/go/chapter05/answer_A30.go
package main

import "fmt"

// Power returns a^b mod m
func Power(a, b, m int) int {
	p := a
	Answer := 1
	for i := 0; i < 30; i++ {
		wari := 1 << uint(i)
		if (b/wari)%2 == 1 {
			Answer = (Answer * p) % m // a の 2^i 乗が掛けられるとき
		}
		p = (p * p) % m
	}
	return Answer
}

// Division returns a/b mod m
func Division(a, b, m int) int {
	return (a * Power(b, m-2, m)) % m
}

func main() {
	var n, r int
	fmt.Scan(&n, &r)
	M := 1000000007

	// 手順 1：分子を求める
	a := 1
	for i := 1; i <= n; i++ {
		a = (a * i) % M
	}

	// 手順 2：分母を求める
	b := 1
	for i := 1; i <= r; i++ {
		b = (b * i) % M
	}
	for i := 1; i <= n-r; i++ {
		b = (b * i) % M
	}

	// 手順 3：答えを求める
	fmt.Println(Division(a, b, M))
}
